### ISSUE TITLE
CreateProcessW FFI binding doesn't match Win32 api from MSDN

### ISSUE THREAD

**Issue by ksubrama** [djberg96/win32-process#19]
```
  attach_pfunc :CreateProcessW,
    [:buffer_in, :buffer_in, :pointer, :pointer, :bool,
     :dword, :buffer_in, :buffer_in, :pointer, :pointer], :bool
```

https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx claims that the second argument (lpCommandLine) is an inout parameter and that the W version of this function can mangle that argument.  I don't know if this is directly causing a bug we're seeing in https://github.com/chef/mixlib-shellout/issues/111 but it seems circumstantially implicated.

Also, I happened to notice that the parameters and return values are :bool instead of BOOL for all the functions/structs in this module.  https://github.com/djberg96/win32-service/issues/27 strikes again!  Only this time, everything has conspired to make things mostly work out - arguments get expanded to 32-bits when pushed on the stack to maintain alignment, Windows seems to mostly return success = 1 and not values >255 (although it is a possibility) and the only struct that had a :bool in it had it at the very end of the struct and structure padding takes care of the remaining bits.

**Comment by djberg96** [comment]
@ksubrama Thanks for the report. I think this should do it: https://github.com/djberg96/win32-process/pull/20

Please review.

**Comment by ksubrama** [comment]
I'm back! :)  I have another followup for this old issue.

https://github.com/djberg96/win32-process/blob/ffi/lib/win32/process/functions.rb#L77 has the same issue with the 6th parameter (lpCommandLine) needing to be an inout.

https://msdn.microsoft.com/en-us/library/windows/desktop/ms682431(v=vs.85).aspx

**Comment by djberg96** [comment]
@ksubrama Thanks for the report. If you wouldn't mind, please do a review of all the functions and see if there's anything else I missed, and then submit a PR. I'll then merge it.

**Comment by tknerr** [comment]
Still getting the `The parameter is incorrect. - CreateProcessW` error with mixlib-shellout 2.2.5 and win32-process 0.8.2

Steps to reproduce:

 * get a Windows 7 64 bit box, using Ruby 2.1.6
 * git clone https://github.com/criteo/cookbook-release
 * cd cookbook-release
 * bundle install
 * bundle exec rake spec

My Ruby:
```
W:\repo\cookbook-release>ruby -v
ruby 2.1.6p336 (2015-04-13 revision 50298) [i386-mingw32]
```

Here's the bundle:
```
W:\repo\cookbook-release>bundle show
Gems included by the bundle:
  * builder (3.2.2)
  * bundler (1.10.6)
  * cookbook-release (0.1)
  * diff-lcs (1.2.5)
  * ffi (1.9.10)
  * highline (1.7.8)
  * mixlib-shellout (2.2.5)
  * rake (10.4.2)
  * rspec (3.4.0)
  * rspec-core (3.4.1)
  * rspec-expectations (3.4.0)
  * rspec-mocks (3.4.0)
  * rspec-support (3.4.1)
  * rspec_junit_formatter (0.2.3)
  * semantic (1.4.1)
  * win32-process (0.8.2)
  * wmi-lite (1.0.0)
```

This is the error:
```
W:\repo\cookbook-release>bundle exec rake spec
W:/tools/chefdk/embedded/bin/ruby.exe -I'W:/home/.chefdk/gem/ruby/2.1.0/gems/rspec-core-3.4.1/lib';'W:/home/.chefdk/gem/ruby/2.1.0/gems/rspec-support-3.4.1/lib' 'W:/home/.chefdk/gem/ruby/2.1.0/gems/rspec-core-3.4.1/exe/rspec' --pattern 'spec/**{,/*/**}/*_spec.rb'
...FFFFF.....

Failures:

  1) GitUtilities.clean_index(?|!) detects clean index
     Failure/Error: cmd.run_command

     SystemCallError:
       The parameter is incorrect. - CreateProcessW
     # W:/home/.chefdk/gem/ruby/2.1.0/gems/mixlib-shellout-2.2.5-universal-mingw32/lib/mixlib/shellout/windows/core_ext.rb:347:in `create'
     # W:/home/.chefdk/gem/ruby/2.1.0/gems/mixlib-shellout-2.2.5-universal-mingw32/lib/mixlib/shellout/windows.rb:80:in `run_command'
     # W:/home/.chefdk/gem/ruby/2.1.0/gems/mixlib-shellout-2.2.5-universal-mingw32/lib/mixlib/shellout.rb:259:in `run_command'
     # ./spec/git_spec.rb:18:in `block (3 levels) in <top (required)>'
     # ./spec/git_spec.rb:16:in `each'
     # ./spec/git_spec.rb:16:in `block (2 levels) in <top (required)>'
...
```

Where the relevant / failing lines in the source specs are [these](https://github.com/criteo/cookbook-release/blob/master/spec/git_spec.rb#L11-L20):
```ruby
    cmds = <<-EOH
    git init .
    git config user.email "you@example.com"
    git config user.name Hello
    EOH
    cmds.split("\n").each do |cmd|
      cmd = Mixlib::ShellOut.new(cmd)
      cmd.run_command   # <------------------ fails here in line 18
      cmd.error!
    end
```

**Comment by ksubrama** [comment]
@djberg96 Sorry I've not had time to review the methods yet.  Will get back to you when I have a PR.

@tknerr Can you open this against mixlib-shellout?  The codepath is related to this library but we have some monkey patches there and I'd like to verify if the issue is upstream or not before we start poking at it here.

**Comment by ksubrama** [comment]
Actually @tknerr, I just chased the problem down.  You need to "cmd.strip" the command before to hand it to Mixlib.  Shellout on unix handles leading spaces.  The Win32 API doesn't appreciate leading spaces in the command name.

**Comment by tknerr** [comment]
Thanks @ksubrama , that was the issue in my case and `cmd.strip` solved it!

**Comment by djberg96** [comment]
@ksubrama Fixed in 0.8.3, which I just released. Thanks for the report!

**Comment by ksubrama** [comment]
Thank you!  And sorry I wasn't able to get to checking the rest of that file sooner.  It's on my list of things to do...
